src/besselj.jl

```diff
@@ -150,13 +150,40 @@ function besselj1(x::Float32)
         return p * s
     end
 end
-"""
-    besselj(nu, x::T) where T <: Union{Float32, Float64}
 
-Bessel function of the first kind of order nu, ``J_{nu}(x)``.
-Nu must be real.
-"""
-function _besselj(nu, x)
+function besselj(nu::Real, x::T) where T
+    abs_nu = abs(nu)
+    abs_x = abs(x)
+
+    Jnu = besselj_positive_args(abs_nu, abs_x)
+
+    if nu >= zero(T)
+        if x >= zero(T)
+            return Jnu
+        else
+            if isinteger(abs_nu)
+                return (-1)^Int(abs_nu) * Jnu
```

Comment:
This shouldn't use `^`.  (rather `iseven(absnu) ? 1 : -1` )

Reply:
Fixed all instances of this. In general there might be better ways to handle this control flow... SpecialFunctions.jl doesn't provide these for negative arguments...

Reply:
Yeah, I'm unsure whether this would be cleaner with a separate method for integer `nu`.

Reply:
Was thinking the same... ideally I could have a `nu::Int` method but I'd probably also have to have a check within the `nu::Real` method if its an integer. Is there a good way to dispatch `nu=5.0` to the `Int` branch? Or just need to have that check manually...

Reply:
have to check manually.

Reply:
```julia
function besselj(nu::Real, x::T) where T
    isinteger(nu) && return besselj(Int(nu), x)
    abs_nu = abs(nu)
    abs_x = abs(x)

    Jnu = besselj_positive_args(abs_nu, abs_x)
    if nu >= zero(T)
        return x >= zero(T) ? Jnu : cispi(abs_nu)*Jnu
    else
        Ynu = bessely_positive_args(abs_nu, abs_x)
        spi, cpi = sincospi(abs_nu)
        out = cpi*Jnu - spi*Ynu
        return x >= zero(T) ? out : out * cispi(nu)
    end
end
function besselj(nu::Int, x::T) where T
    abs_nu = abs(nu)
    abs_x = abs(x)
    sg = (iseven(Int(abs_nu)) ? 1 : -1)

    Jnu = besselj_positive_args(abs_nu, abs_x)
    if nu >= zero(T)
        return x >= zero(T) ? Jnu : Jnu * sg
    else
        Ynu = bessely_positive_args(abs_nu, abs_x)
        spi, cpi = sincospi(abs_nu)
        return (cpi*Jnu - spi*Ynu) * sg
    end
end
```

Reply:
I think i like this better...?

Reply:
yeah. It's not quite correct though since `nu` can be bigger than `2^63` (although I don't know if anyone will call it on numbers that big)

Reply:
I've updated these methods. Hmmm ya these will actually return numbers for such high values so is there an appropriate way to handle this usually? Though you should probably not be using limited precision for such high values....

Reply:
https://github.com/JuliaLang/julia/blob/0b9eda116d88cf472df2d485cfd366bfab17afa8/base/math.jl#L1101 is how I handled it for floating point powers.